File: src/test/java/com/raychenon/leetcode/HIndexTest.java
package com.raychenon.leetcode;

import org.junit.BeforeClass;
import org.junit.Test;

import static junit.framework.TestCase.assertEquals;

/**
 * User: raychenon
 * Date: 11/8/2020
 * https://leetcode.com/problems/h-index/
 */
public class HIndexTest {

    private static HIndex hIndex;

    @BeforeClass
    public static void setUp() {
        hIndex = new HIndex();
    }

    @Test
    public void example1Test() {
        evaluateHIndex(3, new int[]{3, 0, 6, 1, 5});
    }

    @Test
    public void hIndexEqualLengthTest() {
        evaluateHIndex(6, new int[]{12, 7, 8, 9, 10, 11});
    }

    @Test
    public void hIndexValueNotInCitations() {
        evaluateHIndex(5, new int[]{12, 7, 8, 9, 10, 3});
    }


    @Test
    public void only0Test() {
        evaluateHIndex(0, new int[]{0, 0, 0});
    }

    private void evaluateHIndex(int expected, int[] citations) {
        assertEquals(expected, hIndex.hIndex(citations));
        assertEquals(expected, hIndex.hIndexBucketSort(citations));
    }

}

File: src/test/kotlin/com/raychenon/kotlin/leetcode/HIndexTest.kt
package com.raychenon.kotlin.leetcode

import junit.framework.TestCase
import org.junit.Test

/**
 * User: raychenon
 * Date: 11/8/2020
 * https://leetcode.com/problems/h-index/
 */
class HIndexTest {

    @Test
    fun example1Test() {
        evaluateHIndex(3, intArrayOf(3, 0, 6, 1, 5))
    }

    @Test
    fun only0Test() {
        evaluateHIndex(0, intArrayOf(0, 0, 0))
    }

    @Test
    fun hIndexEqualLengthTest() {
        evaluateHIndex(6, intArrayOf(12, 7, 8, 9, 10, 11))
    }

    @Test
    fun hIndexValueNotInCitations() {
        evaluateHIndex(5, intArrayOf(12, 7, 8, 9, 10, 3))
    }

    private fun evaluateHIndex(expected: Int, citations: IntArray) {
        TestCase.assertEquals(expected, HIndex.hIndex(citations))
        TestCase.assertEquals(expected, HIndex.hIndexCountingSort(citations))
    }

}